python/whylogs/core/relations.py

```diff
@@ -0,0 +1,167 @@
+            return x < value
+        if op == Relation.leq:
+            return x <= value
+        if op == Relation.greater:
+            return x > value
+        if op == Relation.geq:
+            return x >= value
+        if op == Relation.neq:
+            return x != value
+        if op == Relation._udf:
+            return self._udf(x)
+        if op == Relation._and:
+            return self._left(x) and self._right(x)
+        if op == Relation._or:
+            return self._left(x) or self._right(x)
+        if op == Relation._not:
+            return not self._right(x)
+
+        raise ValueError("Unknown ConditionCountMetric predicate")
+
+    @classmethod
+    def matches(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.match, value)
+
+    @classmethod
+    def fullmatch(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.fullmatch, value)
+    
+    @classmethod
+    def equals(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.equal, value)
+
+    @classmethod
+    def less_than(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.less, value)
+
+    @classmethod
+    def less_or_equals(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.leq, value)
+
+    @classmethod
+    def greater_than(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.greater, value)
+
+    @classmethod
+    def greater_or_equals(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.geq, value)
+
+    @classmethod
+    def not_equal(cls, value: Union[str, int, float, ValueGetter]) -> "Predicate":
+        return Predicate(Relation.neq, value)
+
+    def and_(self, right: "Predicate") -> "Predicate":
+        return Predicate(Relation._and, left = self, right = right)
+
+    def or_(self, right: "Predicate") -> "Predicate":
+        return Predicate(Relation._or, left = self, right = right)
+
+    @classmethod
+    def is_(cls, udf: Callable[[Any], bool]) -> "Predicate":
+        return Predicate(Relation._udf, udf=udf)
+
+    def serialize(self) -> str:
+        if not (self._left or self._right):
+            return f"{_TOKEN[self._op.value]} x {self._value.serialize()}"
+        if self._op == Relation._not:
+            return f"not {self._right.serialize()}"
+        return f"{_TOKEN[self._op.value]} {self._left.serialize()} {self._right.serialize()}"
+
+    def __str__(self) -> str:
+        if not (self._left or self._right):
+            return f"x {_TOKEN[self._op.value]} {self._value.serialize()}"
+        if self._op == Relation._not:
+            return f"not {str(self._right)}"
+        return f"({str(self._left)}) {_TOKEN[self._op.value]} ({str(self._right)})"
+
+def Not(p: Predicate) -> Predicate:
```

Comment:
maybe `def not_(...)` ?

Reply:
Idk I like the not equal for ease

Reply:
Oh yeah, this does seem a bit weird. hmm..... I think it's ok though. Negate could work better

Reply:
I think some spelling of `not` would be more familiar to programmers. `not_` seems more consistent with `and_` and `or_`. I just did `Not` to be consistent with `X` and not the reserved `not` keyword...

Reply:
Ok, I still think 'negate' would be more clear while keeping the still and not doing the keyword. But I'd be ok with Not